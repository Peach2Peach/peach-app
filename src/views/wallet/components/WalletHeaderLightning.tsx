import { Header } from "../../../components/Header";
import { useStackNavigation } from "../../../hooks/useStackNavigation";
import i18n from "../../../utils/i18n";
import { headerIcons } from "../../../utils/layout/headerIcons";

export const WalletHeaderLightning = () => {
  const navigation = useStackNavigation();
  return (
    <Header
      title={i18n("wallet.lightning.title")}
      icons={[
        {
          ...headerIcons.list,
          accessibilityHint: `${i18n("goTo")} ${i18n("wallet.transactionHistory")}`,
          onPress: () => navigation.navigate("transactionHistoryLightning"),
        },
      ]}
    />
  );
};
