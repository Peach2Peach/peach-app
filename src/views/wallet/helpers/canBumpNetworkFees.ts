import { PeachWallet } from "../../../utils/wallet/PeachWallet";

export const canBumpNetworkFees = (
  peachWallet: PeachWallet,
  transaction: TransactionSummary,
) =>
  !transaction.confirmed &&
  peachWallet.transactions
    .filter((tx) => !tx.confirmationTime?.height)
    .some(({ txid, sent }) => txid === transaction.id && sent > 0);
