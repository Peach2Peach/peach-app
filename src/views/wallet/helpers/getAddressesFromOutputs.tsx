import {
  Transaction as BitcoinTransaction,
  address as bitcoinAddress,
} from "bitcoinjs-lib";
import {
  Transaction as LiquidTransaction,
  address as liquidAddress,
} from "liquidjs-lib";
import { isDefined } from "../../../utils/validation/isDefined";
import { getLiquidNetwork } from "../../../utils/wallet/getLiquidNetwork";
import { getNetwork } from "../../../utils/wallet/getNetwork";

export const scriptToAddress = (script: Buffer, chain: string) => {
  try {
    return chain === "bitcoin"
      ? bitcoinAddress.fromOutputScript(script, getNetwork())
      : liquidAddress.fromOutputScript(script, getLiquidNetwork());
  } catch (e) {
    return undefined;
  }
};

export const getAddressesFromOutputs = ({
  outs,
  chain,
}: Pick<BitcoinTransaction | LiquidTransaction, "outs"> & { chain: Chain }) =>
  outs
    .map((v) => v.script)
    .filter((script) => script.byteLength > 1)
    .map((script) => scriptToAddress(script, chain))
    .filter(isDefined);
