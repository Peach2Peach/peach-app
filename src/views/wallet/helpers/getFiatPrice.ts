import { SATSINBTC } from "../../../constants";
import { round } from "../../../utils/math/round";

export const getFiatPrice = (amount: number, btcPrice: number) =>
  round(((btcPrice || 0) / SATSINBTC) * amount, 2).toFixed(2);
