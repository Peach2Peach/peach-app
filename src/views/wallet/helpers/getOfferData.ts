import { ContractSummary } from "../../../../peach-api/src/@types/contract";

export function getOfferData(
  offers: (SellOffer | BuyOffer)[],
  contracts: ContractSummary[],
  type: TransactionType,
) {
  return offers.map((offer) => {
    const contract = contracts.find((c) => c.id === offer.contractId);
    return {
      offerId: offer.id,
      contractId: offer.contractId,
      amount:
        contract?.amount ||
        (Array.isArray(offer.amount) ? offer.amount[0] : offer.amount),
      address:
        offer.type === "bid"
          ? offer.releaseAddress
          : type === "ESCROWFUNDED"
            ? offer.escrow || ""
            : offer.returnAddress,
      currency: contract?.currency,
      price: contract?.price,
    };
  });
}
