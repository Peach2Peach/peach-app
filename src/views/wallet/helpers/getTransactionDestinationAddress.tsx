import { getHighestValueOutputAddress } from '../../../utils/bitcoin'

const incomingTypes = ['REFUND', 'TRADE', 'DEPOSIT']
export const getTransactionDestinationAddress = (type: TransactionType, transactionDetails: Transaction) => {
  let address

  if (type === 'ESCROWFUNDED') {
    address = transactionDetails.vout.find(
      ({ scriptpubkey_type }) => scriptpubkey_type === 'v0_p2wsh',
    )?.scriptpubkey_address
  } else if (incomingTypes.includes(type)) {
    address = transactionDetails.vout.find(
      ({ scriptpubkey_type }) => scriptpubkey_type === 'v0_p2wpkh',
    )?.scriptpubkey_address
  }
  return address || getHighestValueOutputAddress(transactionDetails)
}
