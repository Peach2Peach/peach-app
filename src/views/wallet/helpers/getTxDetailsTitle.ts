import i18n from '../../../utils/i18n'
import { offerIdToHex } from '../../../utils/offer'

export const getTxDetailsTitle = (tx: TransactionSummary): string => {
  if (tx.offerId) {
    return i18n(tx.type === 'TRADE' ? 'wallet.stackedSats' : 'wallet.refund', offerIdToHex(tx.offerId))
  }
  return i18n(tx.type === 'WITHDRAWAL' ? 'wallet.withdrawal' : 'wallet.deposit')
}
