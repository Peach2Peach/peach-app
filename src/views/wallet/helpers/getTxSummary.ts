import { TransactionDetails } from 'bdk-rn/lib/classes/Bindings'
import { bitcoinStore } from '../../../store/bitcoinStore'
import { tradeSummaryStore } from '../../../store/tradeSummaryStore'
import { getTransactionType, txIsConfirmed } from '../../../utils/transaction'
import { walletStore } from '../../../utils/wallet/walletStore'

export const getTxSummary = (tx: TransactionDetails) => {
  const offerId = walletStore.getState().txOfferMap[tx.txid]
  const offer = tradeSummaryStore.getState().getOffer(offerId)
  const sats = Math.abs(tx.received - tx.sent)
  const price = sats / bitcoinStore.getState().satsPerUnit
  const type = getTransactionType(tx, offer)

  return {
    id: tx.txid,
    offerId,
    contractId: offer?.contractId,
    type,
    amount: sats,
    price,
    currency: bitcoinStore.getState().currency,
    date: txIsConfirmed(tx) ? new Date((tx.confirmationTime?.timestamp || Date.now()) * 1000) : new Date(),
    confirmed: txIsConfirmed(tx),
  }
}
