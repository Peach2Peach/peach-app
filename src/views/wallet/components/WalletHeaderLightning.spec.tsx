import { fireEvent, render } from "test-utils";
import { navigateMock } from "../../../../tests/unit/helpers/NavigationWrapper";
import { WalletHeaderLightning } from "./WalletHeaderLightning";

describe("WalletHeaderLightning", () => {
  it("should render correctly", () => {
    const { toJSON } = render(<WalletHeaderLightning />);
    expect(toJSON()).toMatchSnapshot();
  });
  it("should go to transaction history", () => {
    const { getByAccessibilityHint } = render(<WalletHeaderLightning />);

    fireEvent.press(getByAccessibilityHint("go to transaction history"));
    expect(navigateMock).toHaveBeenCalledWith("transactionHistoryLightning");
  });
});
