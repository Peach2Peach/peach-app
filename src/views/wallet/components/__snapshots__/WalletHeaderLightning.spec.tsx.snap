// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletHeaderLightning should render correctly 1`] = `
<View
  style={
    [
      {
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderBottomWidth": 1,
      },
      {
        "paddingTop": 0,
        "zIndex": 1,
      },
      {
        "backgroundColor": "#FFF9F6",
      },
      {
        "borderColor": "#FEEDE5",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 8,
          "paddingBottom": 6,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 6,
        },
        {},
        undefined,
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexBasis": "0%",
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "gap": 4,
        }
      }
    >
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          [
            {
              "color": "#2B1911",
              "fontFamily": "Baloo2-Regular",
              "fontSize": 16,
              "letterSpacing": 0.16,
              "lineHeight": 24,
            },
            [
              {
                "fontFamily": "Baloo2-Bold",
                "fontSize": 16,
                "letterSpacing": 0.16,
                "lineHeight": 26,
              },
              {},
              {
                "color": "#2B1911",
              },
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ],
            false,
          ]
        }
      >
        Peach wallet
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 10,
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessibilityHint="go to transaction history"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "paddingBottom": 2,
            "paddingLeft": 2,
            "paddingRight": 2,
            "paddingTop": 2,
          }
        }
      >
        <test-file-stub
          fill="#7D675E"
          style={
            [
              {
                "height": 24,
                "width": 24,
              },
              {
                "height": 20,
                "width": 20,
              },
              {},
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
